### Eclipse Workspace Patch 1.0
#P Rakuun_live
Index: source/intern/gui/panel/board/postingview.tpl
===================================================================
--- source/intern/gui/panel/board/postingview.tpl	(revision 429)
+++ source/intern/gui/panel/board/postingview.tpl	(working copy)
@@ -1,13 +1,6 @@
-<? if($this->hasErrors()): ?>
-	<? $this->displayErrors() ?>
-<? else: ?>
-	<h2><?= $this->params->boardname; ?></h2>
-	<ul>
-		<? foreach ($this->panels as $panel): ?>
-			<? if (preg_match('=^posting_(\d)+$=', $panel->getName())): ?>
-				<? $panel->display(); ?>
-			<? endif; ?>
-		<? endforeach; ?>
-	</ul>	
+<? if ($this->hasMessages()): ?>
+	<? $this->displayMessages(); ?>
 <? endif; ?>
-<? $this->displayPanel('addposting'); ?>
\ No newline at end of file
+<h2><?= $this->params->boardname; ?></h2>
+<? $this->displayPanel('board'); ?>
+<? $this->displayPanel('post'); ?>
\ No newline at end of file
Index: source/intern/gui/panel/alliance/board.php
===================================================================
--- source/intern/gui/panel/alliance/board.php	(revision 0)
+++ source/intern/gui/panel/alliance/board.php	(revision 0)
@@ -0,0 +1,35 @@
+<?php
+
+class Rakuun_Intern_GUI_Panel_Alliance_Board extends Rakuun_Intern_GUI_Panel_Board {
+	public function __construct($name, $title = '') {
+		parent::__construct($name, Rakuun_DB_Containers::getBoardsAllianceContainer()->getFilteredContainer(self::getAllianceFilter()), $title);
+	}
+	
+	protected function getFilteredRecord() {
+		$record = new DB_Record();
+		$record->alliance = Rakuun_User_Manager::getCurrentUser()->alliance;
+		return $record;
+	}
+	
+	private static function getAllianceFilter() {
+		$options = array();
+		$options['conditions'][] = array('alliance = ?', Rakuun_User_Manager::getCurrentUser()->alliance);
+		return $options; 
+	}
+	
+	protected function getPostingsContainer() {
+		return Rakuun_DB_Containers::getBoardsAlliancePostingsContainer();
+	}
+	
+	protected function getVisitedContainer() {
+		return Rakuun_DB_Containers::getBoardsAllianceLastVisitedContainer();
+	}
+	
+	public static function getNewPostingsCount(DB_Container $null = null, DB_Container $anothernull = null) {
+		return parent::getNewPostingsCount(
+			Rakuun_DB_Containers::getBoardsAllianceContainer()->getFilteredContainer(self::getAllianceFilter()),
+			Rakuun_DB_Containers::getBoardsAllianceLastVisitedContainer()
+		);
+	}
+}
+?>
\ No newline at end of file
Index: source/intern/gui/panel/board/addposting.tpl
===================================================================
--- source/intern/gui/panel/board/addposting.tpl	(revision 429)
+++ source/intern/gui/panel/board/addposting.tpl	(working copy)
@@ -1,7 +0,0 @@
-<? if($this->hasErrors()): ?>
-	<? $this->displayErrors() ?>
-<? endif ?>
-<hr />
-<? $this->displayLabelForPanel('text') ?> <? $this->displayPanel('text') ?>
-<br class="clear" />
-<? $this->displayPanel('submit') ?>
\ No newline at end of file
Index: source/intern/gui/panel/board/board.tpl
===================================================================
--- source/intern/gui/panel/board/board.tpl	(revision 450)
+++ source/intern/gui/panel/board/board.tpl	(working copy)
@@ -1,12 +1,19 @@
-<? if($this->hasErrors()): ?>
-	<? $this->displayErrors() ?>
-<? endif ?>
-<tr>
-	<td><? $this->displayPanel('boardlink'); ?></td>
-	<td class="posting_count"><?= $this->params->count; ?></td>
-	<td>
-		<? if ($this->hasPanel('date')): ?>
-			<? $this->displayPanel('date'); ?>
-		<? endif; ?>
-	</td>
-</tr>
\ No newline at end of file
+<? if ($this->hasMessages()): ?>
+	<? $this->displayMessages(); ?>
+<? endif; ?>
+
+<? if ($this->hasPanel('markread')): ?>
+	<? $this->displayPanel('markread'); ?>
+<? endif; ?>
+<? $this->displayPanel('board'); ?>
+<? if ($this->getPageCount() > 1): ?>
+	<? $this->displayLabelForPanel('pages'); ?>: <? $this->displayPanel('pages'); ?>
+	<hr />
+<? endif; ?>
+<hr />
+<? if ($this->hasPanel('name')): ?>
+	<? $this->displayPanel('name'); ?>
+<? endif; ?>
+<? if ($this->hasPanel('addboard')): ?>
+	<? $this->displayPanel('addboard'); ?>
+<? endif; ?>
\ No newline at end of file
Index: source/db/containers.php
===================================================================
--- source/db/containers.php	(revision 544)
+++ source/db/containers.php	(working copy)
@@ -27,6 +27,12 @@
 	private static $boardsContainer = null;
 	private static $boardsPostingsContainer = null;
 	private static $boardsLastVisitedContainer = null;
+	private static $boardsAllianceContainer = null;
+	private static $boardsAlliancePostingsContainer = null;
+	private static $boardsAllianceLastVisitedContainer = null;
+	private static $boardsMetaContainer = null;
+	private static $boardsMetaPostingsContainer = null;
+	private static $boardsMetaLastVisitedContainer = null;
 	private static $alliancesDiplomaciesContainer = null;
 	private static $metasContainer = null;
 	private static $metasApplicationsContainer = null;
@@ -303,30 +309,6 @@
 	/**
 	 * @return DB_Container
 	 */
-	public static function getBoardsContainer() {
-		if (self::$boardsContainer)
-			return self::$boardsContainer;
-			
-		self::$boardsContainer = new DB_Container('boards');
-		
-		return self::$boardsContainer;
-	}
-	
-	/**
-	 * @return DB_Container
-	 */
-	public static function getBoardsPostingsContainer() {
-		if (self::$boardsPostingsContainer)
-			return self::$boardsPostingsContainer;
-			
-		self::$boardsPostingsContainer = new DB_Container('boards_postings');
-		
-		return self::$boardsPostingsContainer;
-	}
-	
-	/**
-	 * @return DB_Container
-	 */
 	public static function getAlliancesDiplomaciesContainer() {
 		if (self::$alliancesDiplomaciesContainer)
 			return self::$alliancesDiplomaciesContainer;
@@ -799,6 +781,30 @@
 	/**
 	 * @return DB_Container
 	 */
+	public static function getBoardsContainer() {
+		if (self::$boardsContainer)
+			return self::$boardsContainer;
+			
+		self::$boardsContainer = new DB_Container('boards');
+		
+		return self::$boardsContainer;
+	}
+	
+	/**
+	 * @return DB_Container
+	 */
+	public static function getBoardsPostingsContainer() {
+		if (self::$boardsPostingsContainer)
+			return self::$boardsPostingsContainer;
+			
+		self::$boardsPostingsContainer = new DB_Container('boards_postings');
+		
+		return self::$boardsPostingsContainer;
+	}
+	
+	/**
+	 * @return DB_Container
+	 */
 	public static function getBoardsLastVisitedContainer() {
 		if (self::$boardsLastVisitedContainer)
 			return self::$boardsLastVisitedContainer;
@@ -808,6 +814,80 @@
 		
 		return self::$boardsLastVisitedContainer;
 	}
+	
+	/**
+	 * @return DB_Container
+	 */
+	public static function getBoardsAllianceContainer() {
+		if (self::$boardsAllianceContainer)
+			return self::$boardsAllianceContainer;
+			
+		self::$boardsAllianceContainer = new DB_Container('boards_alliance');
+		
+		return self::$boardsAllianceContainer;
+	}
+	
+	/**
+	 * @return DB_Container
+	 */
+	public static function getBoardsAlliancePostingsContainer() {
+		if (self::$boardsAlliancePostingsContainer)
+			return self::$boardsAlliancePostingsContainer;
+			
+		self::$boardsAlliancePostingsContainer = new DB_Container('boards_alliance_postings');
+		
+		return self::$boardsAlliancePostingsContainer;
+	}
+	
+	/**
+	 * @return DB_Container
+	 */
+	public static function getBoardsAllianceLastVisitedContainer() {
+		if (self::$boardsAllianceLastVisitedContainer)
+			return self::$boardsAllianceLastVisitedContainer;
+		
+		self::$boardsAllianceLastVisitedContainer = new DB_Container('boards_alliance_visited');
+		self::$boardsAllianceLastVisitedContainer->addReferencedContainer(self::getBoardsAllianceContainer());
+		
+		return self::$boardsAllianceLastVisitedContainer;
+	}
+	
+	/**
+	 * @return DB_Container
+	 */
+	public static function getBoardsMetaContainer() {
+		if (self::$boardsMetaContainer)
+			return self::$boardsMetaContainer;
+			
+		self::$boardsMetaContainer = new DB_Container('boards_meta');
+		
+		return self::$boardsMetaContainer;
+	}
+	
+	/**
+	 * @return DB_Container
+	 */
+	public static function getBoardsMetaPostingsContainer() {
+		if (self::$boardsMetaPostingsContainer)
+			return self::$boardsMetaPostingsContainer;
+			
+		self::$boardsMetaPostingsContainer = new DB_Container('boards_meta_postings');
+		
+		return self::$boardsMetaPostingsContainer;
+	}
+	
+	/**
+	 * @return DB_Container
+	 */
+	public static function getBoardsMetaLastVisitedContainer() {
+		if (self::$boardsMetaLastVisitedContainer)
+			return self::$boardsMetaLastVisitedContainer;
+		
+		self::$boardsMetaLastVisitedContainer = new DB_Container('boards_meta_visited');
+		self::$boardsMetaLastVisitedContainer->addReferencedContainer(self::getBoardsMetaContainer());
+		
+		return self::$boardsMetaLastVisitedContainer;
+	}
 }
 
 ?>
\ No newline at end of file
Index: source/intern/gui/panel/board/boardview.php
===================================================================
--- source/intern/gui/panel/board/boardview.php	(revision 546)
+++ source/intern/gui/panel/board/boardview.php	(working copy)
@@ -1,169 +0,0 @@
-<?php
-
-/**
- * Displays all boards.
- */
-class Rakuun_Intern_GUI_Panel_Board_Boardview extends GUI_Panel {
-	/**
-	 * HINT: to add another boardtype you have to go through this points:
-	 * - add new const TYPE_...
-	 * - edit self::showBoard()
-	 * - edit self::showForum()
-	 * - edit Rakuun_Intern_GUI_Panel_Board_Addboard::onSubmit()
-	 */	
-	const TYPE_ALLIANCE = 1;
-	const TYPE_META = 2;
-	const TYPE_ADMIN = 3;
-	
-	protected $current = null;
-	private $boardtype = null;
-	
-	public function __construct($name, $type, $title = '') {
-		$this->boardtype = $type;
-		parent::__construct($name, $title);
-	}
-	
-	public function init() {
-		parent::init();
-		$this->current = Router::get()->getCurrentModule();
-		if (($boardId = $this->current->getParam('board')) > 0) {
-			// Board darstellen
-			$this->showBoard($boardId);
-		} else {
-			$this->showForum();
-		}
-	}
-	
-	protected function showBoard($boardId) {
-		$this->setTemplate(dirname(__FILE__).'/boardview.tpl');
-		switch ($this->boardtype) {
-			case self::TYPE_ALLIANCE: 
-				$options['conditions'][] = array('alliance = ?', Rakuun_User_Manager::getCurrentUser()->alliance);
-				$options['conditions'][] = array('type = ?', self::TYPE_ALLIANCE);
-				$board = Rakuun_DB_Containers::getBoardsContainer()->selectByIdFirst($boardId, $options);
-			break;
-			case self::TYPE_META:
-				$options['conditions'][] = array('meta = ?', Rakuun_User_Manager::getCurrentUser()->alliance->meta);
-				$options['conditions'][] = array('type = ?', self::TYPE_META);
-				$board = Rakuun_DB_Containers::getBoardsContainer()->selectByIdFirst($boardId, $options);
-			break;
-			case self::TYPE_ADMIN:
-				$options['conditions'][] = array('type = ?', self::TYPE_ADMIN);
-				$board = Rakuun_DB_Containers::getBoardsContainer()->selectByIdFirst($boardId, $options);
-			break;
-			default:
-				$this->addError('Unbekannter Boardtype: '.$this->boardtype);
-				return;
-			break;
-		}
-		if (!$board) {
-			$this->addError('Du hast keinen Zugriff auf das Forum mit der ID '.$boardId);
-			return;
-		}
-		$this->addPanel(new Rakuun_Intern_GUI_Panel_Board_Postingview('board', $board));
-		$this->params->boardname = 'board';
-	}
-	
-	protected function showForum() {		
-		$this->setTemplate(dirname(__FILE__).'/boardview.tpl');
-		Rakuun_DB_Containers::getBoardsContainer()->addInsertCallback(array($this, 'onAddBoard'));
-		switch ($this->boardtype) {
-			case self::TYPE_ALLIANCE: 
-				$boards = Rakuun_DB_Containers::getBoardsContainer()->selectByAlliance(Rakuun_User_Manager::getCurrentUser()->alliance);
-			break;
-			case self::TYPE_META:
-				$boards = Rakuun_DB_Containers::getBoardsContainer()->selectByMeta(Rakuun_User_Manager::getCurrentUser()->alliance->meta);
-			break;
-			case self::TYPE_ADMIN: 
-				$boards = Rakuun_DB_Containers::getBoardsContainer()->selectByType(self::TYPE_ADMIN);
-			break;
-			default:
-				$this->addError('Fehlerhafter Boardtype: '.$this->boardtype);
-				return;
-			break;
-		}		
-		$postings = Rakuun_DB_Containers::getBoardsPostingsContainer()->select(array('order' => 'date DESC'));//, 'group' => 'board'));
-		if ($postings && $boards) {
-			$order = new DB_Order($boards, $postings);
-			$boards = $order->desc();
-		}
-		if (is_array($boards)) {
-			foreach ($boards as $board) {
-				$this->onAddBoard($board);
-			}
-		}
-		$this->addPanel($blanko = new GUI_Panel('markread'));
-		$blanko->addPanel(new GUI_Control_SubmitButton('markread', 'Alle als gelesen markieren'));
-		$this->addPanel(new Rakuun_Intern_GUI_Panel_Board_Addboard('addboard', $this->boardtype), 'Board hinzufügen');
-	}
-	
-	public static function getNewPostingsCount($boardtype = 0) {
-		$user = Rakuun_User_Manager::getCurrentUser();
-		$boardsContainer = Rakuun_DB_Containers::getBoardsContainer();
-		$postingsContainer = Rakuun_DB_Containers::getBoardsPostingsContainer();
-		$lastVisitedContainer = Rakuun_DB_Containers::getBoardsLastVisitedContainer();
-		$boards = array();
-		switch ($boardtype) {
-			case self::TYPE_ALLIANCE:
-				$boards = $boardsContainer->selectByType(self::TYPE_ALLIANCE, array('conditions' => array(array('alliance = ?', $user->alliance))));
-			break;
-			case self::TYPE_META:
-				$boards = $boardsContainer->selectByType(self::TYPE_META, array('conditions' => array(array('meta = ?', $user->alliance->meta))));
-			break;
-			case self::TYPE_ADMIN:
-				$boards = $boardsContainer->selectByType(self::TYPE_ADMIN);
-			break;
-		}
-		$count = 0;
-		foreach ($boards as $board) {
-			$options = array();
-			$options['order'] = 'date DESC';
-			$options['conditions'][] = array('board = ?', $board);
-			$visitedOptions = array();
-			$visitedOptions['conditions'][] = array('board = ?', $board);
-			$visitedOptions['conditions'][] = array('user = ?', $user);
-			$lastVisited = $lastVisitedContainer->selectFirst($visitedOptions);
-			$posting = $postingsContainer->selectFirst($options);
-			if ($lastVisited === null || ($posting && $posting->date > $lastVisited->date)) {
-				$count++;
-			}
-		}
-		return $count;
-	}
-	
-	public function onMarkread() {
-		$user = Rakuun_User_Manager::getCurrentUser();
-		$boardsContainer = Rakuun_DB_Containers::getBoardsContainer();
-		$lastVisitedContainer = Rakuun_DB_Containers::getBoardsLastVisitedContainer();
-		$boards = array();
-		switch ($this->boardtype) {
-			case self::TYPE_ALLIANCE:
-				$boards = $boardsContainer->selectByType(self::TYPE_ALLIANCE, array('conditions' => array(array('alliance = ?', $user->alliance))));
-			break;
-			case self::TYPE_META:
-				$boards = $boardsContainer->selectByType(self::TYPE_META, array('conditions' => array(array('meta = ?', $user->alliance->meta))));
-			break;
-			case self::TYPE_ADMIN:
-				$boards = $boardsContainer->selectByType(self::TYPE_ADMIN);
-			break;
-		}
-		DB_Connection::get()->beginTransaction();
-		foreach ($boards as $board) {
-			$lastVisitedContainer->deleteByBoard($board);
-			$newVisit = new DB_Record();
-			$newVisit->board = $board;
-			$newVisit->user = $user;
-			$newVisit->date = time();
-			$lastVisitedContainer->save($newVisit);
-		}
-		DB_Connection::get()->commit();
-		$this->getModule()->invalidate();
-	}
-	
-	// CALLBACKS ---------------------------------------------------------------
-	public function onAddBoard(DB_Record $board) {
-		$this->addPanel(new Rakuun_Intern_GUI_Panel_Board_Board('board_'.$board->getPK(), $board));
-	}
-}
-
-?>
\ No newline at end of file
Index: source/intern/gui/panel/admin/board.php
===================================================================
--- source/intern/gui/panel/admin/board.php	(revision 429)
+++ source/intern/gui/panel/admin/board.php	(working copy)
@@ -6,7 +6,7 @@
 		
 		$this->setTemplate(dirname(__FILE__).'/board.tpl');
 		
-		$this->addPanel(new Rakuun_Intern_GUI_Panel_Board_Boardview('board', Rakuun_Intern_GUI_Panel_Board_Boardview::TYPE_ADMIN), 'Admin Forum');
+		$this->addPanel(new Rakuun_Intern_GUI_Panel_Board('board'), 'Admin Forum');
 	}	
 }
 
Index: source/intern/gui/panel/board/addboard.tpl
===================================================================
--- source/intern/gui/panel/board/addboard.tpl	(revision 429)
+++ source/intern/gui/panel/board/addboard.tpl	(working copy)
@@ -1,6 +0,0 @@
-<? if($this->hasErrors()): ?>
-	<? $this->displayErrors() ?>
-<? endif ?>
-<? $this->displayLabelForPanel('name') ?> <? $this->displayPanel('name') ?>
-<br class="clear" />
-<? $this->displayPanel('submit') ?>
\ No newline at end of file
Index: source/intern/module/meta/navigation.php
===================================================================
--- source/intern/module/meta/navigation.php	(revision 429)
+++ source/intern/module/meta/navigation.php	(working copy)
@@ -16,7 +16,7 @@
 		if ($parent->hasSubmodule('kick'))
 			$navigation->addModuleNode($parent->getSubmodule('kick'), 'Allianz kicken');
 		if ($parent->hasSubmodule('board'))
-			$navigation->addModuleNode($parent->getSubmodule('board'), 'Forum ('.Rakuun_Intern_GUI_Panel_Board_Boardview::getNewPostingsCount(Rakuun_Intern_GUI_Panel_Board_Boardview::TYPE_META).')');
+			$navigation->addModuleNode($parent->getSubmodule('board'), 'Forum ('.Rakuun_Intern_GUI_Panel_Meta_Board::getNewPostingsCount().')');
 		if ($parent->hasSubmodule('statistics'))
 			$navigation->addModuleNode($parent->getSubmodule('statistics'), 'Statistiken');
 		if ($parent->hasSubmodule('polls'))
Index: source/intern/module/meta/board.php
===================================================================
--- source/intern/module/meta/board.php	(revision 429)
+++ source/intern/module/meta/board.php	(working copy)
@@ -7,7 +7,7 @@
 		$this->setPageTitle('Forum - '.$this->getUser()->alliance->meta->name);
 		$this->contentPanel->setTemplate(dirname(__FILE__).'/board.tpl');
 		
-		$this->contentPanel->addPanel(new Rakuun_GUI_Panel_Box('board', new Rakuun_Intern_GUI_Panel_Board_Boardview('Forum', Rakuun_Intern_GUI_Panel_Board_Boardview::TYPE_META), 'Meta Forum'));
+		$this->contentPanel->addPanel(new Rakuun_GUI_Panel_Box('board', new Rakuun_Intern_GUI_Panel_Meta_Board('Forum'), 'Meta Forum'));
 	}
 }
 
Index: source/intern/gui/panel/board/posting.php
===================================================================
--- source/intern/gui/panel/board/posting.php	(revision 497)
+++ source/intern/gui/panel/board/posting.php	(working copy)
@@ -7,20 +7,21 @@
 	private $posting = null;
 	
 	public function __construct($name, DB_Record $posting) {
+		parent::__construct($name);
 		$this->posting = $posting;
-		parent::__construct($name);
 	}
 	
 	public function init() {
 		parent::init();
-		
+		 
 		$this->setTemplate(dirname(__FILE__).'/posting.tpl');
 		if (Router::get()->getCurrentModule()->getParam('edit') == $this->posting->getPK()
 			&& Rakuun_User_Manager::getCurrentUser()->getPK() == $this->posting->user->getPK()
 		) {
-			$this->addPanel($text = new GUI_Control_TextArea('text', $this->posting->text, 'Posting'));
+			$this->addPanel($blanko = new GUI_Panel('form'));
+			$blanko->addPanel($text = new GUI_Control_TextArea('text', $this->posting->text, 'Posting'));
 			$text->addValidator(new GUI_Validator_Mandatory());
-			$this->addPanel(new GUI_Control_SubmitButton('submit', 'speichern'));
+			$blanko->addPanel(new GUI_Control_SubmitButton('edit', 'speichern'));
 		}
 		$this->params->posting = $this->posting;
 		$this->addPanel(new GUI_Panel_Date('date', $this->posting->date));
@@ -32,13 +33,14 @@
 		}
 	}
 	
-	public function onSubmit() {
+	public function onEdit() {
 		if ($this->hasErrors())
 			return;
 		
-		$this->posting->text = $this->text->getValue();
+		$this->posting->text = $this->form->text->getValue();
 		$this->posting->editdate = time();
-		Rakuun_DB_Containers::getBoardsPostingsContainer()->save($this->posting);
+		$this->posting->save();
+		$this->getModule()->redirect($this->getModule()->getUrl(array('board' => $this->posting->board)));
 	}
 }
 
Index: source/intern/module/alliance/navigation.php
===================================================================
--- source/intern/module/alliance/navigation.php	(revision 429)
+++ source/intern/module/alliance/navigation.php	(working copy)
@@ -26,7 +26,7 @@
 		if ($parent->hasSubmodule('diplomacy'))
 			$navigation->addModuleNode($parent->getSubmodule('diplomacy'), 'Diplomatie');
 		if ($parent->hasSubmodule('board'))
-			$navigation->addModuleNode($parent->getSubmodule('board'), 'Forum ('.Rakuun_Intern_GUI_Panel_Board_Boardview::getNewPostingsCount(Rakuun_Intern_GUI_Panel_Board_Boardview::TYPE_ALLIANCE).')');
+			$navigation->addModuleNode($parent->getSubmodule('board'), 'Forum ('.Rakuun_Intern_GUI_Panel_Alliance_Board::getNewPostingsCount().')');
 		if ($parent->hasSubmodule('statistics'))
 			$navigation->addModuleNode($parent->getSubmodule('statistics'), 'Statistiken');
 		if ($parent->hasSubmodule('polls'))
Index: source/intern/gui/panel/board/postingview.php
===================================================================
--- source/intern/gui/panel/board/postingview.php	(revision 429)
+++ source/intern/gui/panel/board/postingview.php	(working copy)
@@ -3,41 +3,65 @@
 /**
  * Displays all threads of a specific board.
  */
-class Rakuun_Intern_GUI_Panel_Board_Postingview extends GUI_Panel {
+class Rakuun_Intern_GUI_Panel_Board_PostingView extends GUI_Panel {
 	private $board = null;
+	private $postingsContainer = null;
+	private $visitedContainer = null;
 	
-	public function __construct($name, DB_Record $board) {
+	public function __construct($name, DB_Record $board, DB_Container $postingsContainer, DB_Container $visitedContainer) {
+		parent::__construct($name);
+		
 		$this->board = $board;
+		$this->postingsContainer = $postingsContainer;
+		$this->visitedContainer = $visitedContainer;
+	}
+	
+	public function init() {
+		parent::init();
+		
+		$this->setTemplate(dirname(__FILE__).'/postingview.tpl');
 		$options = array();
 		$options['conditions'][] = array('user = ?', Rakuun_User_Manager::getCurrentUser());
-		$options['conditions'][] = array('board = ?', $board);
-		$lastVisit = Rakuun_DB_Containers::getBoardsLastVisitedContainer()->selectFirst($options);
+		$options['conditions'][] = array('board = ?', $this->board);
+		$lastVisit = $this->visitedContainer->selectFirst($options);
 		if (!$lastVisit)
 			$lastVisit = new DB_Record();
 		$lastVisit->user = Rakuun_User_Manager::getCurrentUser();
-		$lastVisit->board = $board;
+		$lastVisit->board = $this->board;
 		$lastVisit->date = time();
-		Rakuun_DB_Containers::getBoardsLastVisitedContainer()->save($lastVisit);
-		
-		parent::__construct($name);		
-	}
-	
-	public function init() {
-		parent::init();
+		$this->visitedContainer->save($lastVisit);
 		
-		$this->setTemplate(dirname(__FILE__).'/postingview.tpl');
-		Rakuun_DB_Containers::getBoardsPostingsContainer()->addInsertCallback(array($this, 'onAddBoardPosting'));		
-		$postings = Rakuun_DB_Containers::getBoardsPostingsContainer()->selectByBoard($this->board, array('order' => 'date ASC'));
+		$options = array();
+		$options['conditions'][] = array('board = ?', $this->board);
+		$options['order'] = 'date ASC';
+		$postings = $this->postingsContainer->select($options);
+		$this->addPanel($list = new GUI_Panel_List('board'));
 		foreach ($postings as $posting) {
-			$this->onAddBoardPosting($posting);			
+			$list->addItem($item = new Rakuun_Intern_GUI_Panel_Board_Posting('posting'.$posting->getPK(), $posting));
 		}
-		$this->addPanel(new Rakuun_Intern_GUI_Panel_Board_Addposting('addposting', $this->board));
 		$this->params->boardname = $this->board->name;
+		$this->addPanel($blanko = new GUI_Panel('post'));
+		$blanko->addPanel($text = new GUI_Control_TextArea('text', null, 'Posting'));
+		$text->addValidator(new GUI_Validator_Mandatory());
+		$blanko->addPanel(new GUI_Control_SubmitButton('submit', 'posten'));
 	}
 	
-	public function onAddBoardPosting(DB_Record $posting) {
-		$this->addPanel(new Rakuun_Intern_GUI_Panel_Board_Posting('posting_'.$posting->getPK(), $posting));
+	public function onSubmit() {
+		if ($this->hasErrors())
+			return;
+		
+		DB_Connection::get()->beginTransaction();
+		$posting = new DB_Record();
+		$posting->board = $this->board;
+		$posting->user = Rakuun_User_Manager::getCurrentUser();
+		$posting->text = $this->post->text;
+		$posting->date = time();
+		$this->postingsContainer->save($posting);
+		$this->board->date = time();
+		$this->board->save();
+		DB_Connection::get()->commit();
+		$this->post->text->resetValue();
+		$this->getModule()->invalidate();
 	}
 }
-
 ?>
\ No newline at end of file
Index: source/intern/gui/panel/board/posting.tpl
===================================================================
--- source/intern/gui/panel/board/posting.tpl	(revision 446)
+++ source/intern/gui/panel/board/posting.tpl	(working copy)
@@ -1,31 +1,22 @@
-<? if($this->hasErrors()): ?>
-	<li>
-		<? $this->displayErrors(); ?>
-	</li>
-<? endif ?>
-<li>
-	<div class="rakuun_board_postinfo">
-		<? $this->displayPanel('date'); ?><br />
-		<? $this->displayPanel('user'); ?>:
-		<? if ($this->hasPanel('editlink')): ?>
-			<? $this->displayPanel('editlink'); ?>
-		<? endif; ?>
+<div class="rakuun_board_postinfo">
+	<? $this->displayPanel('date'); ?><br />
+	<? $this->displayPanel('user'); ?>:
+	<? if ($this->hasPanel('editlink')): ?>
+		<? $this->displayPanel('editlink'); ?>
+	<? endif; ?>
+	<br />
+</div>
+<div class="rakuun_board_postcontent">
+	<? if ($this->hasPanel('form')): ?>
+		<? $this->displayPanel('form'); ?>
+	<? else: ?>
+		<?=  Text::format(Text::escapeHTML($this->params->posting->text)); ?>
+		<? if ($this->hasPanel('editdate')): ?>
 		<br />
-	</div>
-	<div class="rakuun_board_postcontent">
-		<? if ($this->hasPanel('submit')): ?>
-			<? $this->displayPanel('text'); ?>
-			<br class="clear" />
-			<? $this->displayPanel('submit'); ?>
-		<? else: ?>
-			<?=  Text::format(Text::escapeHTML($this->params->posting->text)); ?>
-			<? if ($this->hasPanel('editdate')): ?>
-			<br />
-			<i>
-				last edit:
-				<? $this->displayPanel('editdate'); ?>
-			</i>
-			<? endif; ?>
+		<i>
+			last edit:
+			<? $this->displayPanel('editdate'); ?>
+		</i>
 		<? endif; ?>
-	</div>
-</li>
\ No newline at end of file
+	<? endif; ?>
+</div>
\ No newline at end of file
Index: source/intern/gui/panel/board/addboard.php
===================================================================
--- source/intern/gui/panel/board/addboard.php	(revision 429)
+++ source/intern/gui/panel/board/addboard.php	(working copy)
@@ -1,46 +0,0 @@
-<?php
-
-/**
- * Adds a new board to board.
- */
-class Rakuun_Intern_GUI_Panel_Board_Addboard extends GUI_Panel {
-	private $boardtype = null;
-	public function __construct($name, $type) {
-		$this->boardtype = $type;
-		parent::__construct($name);
-	}
-	
-	public function init() {
-		parent::init();
-		
-		$this->setTemplate(dirname(__FILE__).'/addboard.tpl');
-		$this->addPanel($name = new GUI_Control_TextBox('name', null, 'Boardname'));
-		$name->addValidator(new GUI_Validator_Mandatory());
-		$name->addValidator(new GUI_Validator_RangeLength(2, 25));
-		$this->addPanel(new GUI_Control_SubmitButton('submit', 'anlegen'));
-	}
-	
-	public function onSubmit() {
-		$board = new DB_Record();
-		switch ($this->boardtype) {
-			case Rakuun_Intern_GUI_Panel_Board_Boardview::TYPE_ALLIANCE:
-				$board->alliance = Rakuun_User_Manager::getCurrentUser()->alliance;
-			break;
-			case Rakuun_Intern_GUI_Panel_Board_Boardview::TYPE_META:
-				$board->meta = Rakuun_User_Manager::getCurrentUser()->alliance->meta;
-			break;
-			case Rakuun_Intern_GUI_Panel_Board_Boardview::TYPE_ADMIN:
-				//do nothing
-			break;
-			default:
-				$this->addError('Ubekannter Boardtype: '.$this->boardtype);
-				return;
-			break;
-		}
-		$board->name = $this->name;
-		$board->type = $this->boardtype;
-		Rakuun_DB_Containers::getBoardsContainer()->save($board);
-	}
-}
-
-?>
\ No newline at end of file
Index: source/intern/gui/panel/board/addposting.php
===================================================================
--- source/intern/gui/panel/board/addposting.php	(revision 446)
+++ source/intern/gui/panel/board/addposting.php	(working copy)
@@ -1,36 +0,0 @@
-<?php
-
-/**
- * Adds a new posting to a specific board.
- */
-class Rakuun_Intern_GUI_Panel_Board_Addposting extends GUI_Panel {
-	private $board = null;
-	public function __construct($name, $board) {
-		$this->board = $board;
-		parent::__construct($name);
-	}
-	
-	public function init() {
-		parent::init();
-		
-		$this->setTemplate(dirname(__FILE__).'/addposting.tpl');
-		$this->addPanel($text = new GUI_Control_TextArea('text', null, 'Posting'));
-		$text->addValidator(new GUI_Validator_Mandatory());
-		$this->addPanel(new GUI_Control_SubmitButton('submit', 'posten'));
-	}
-	
-	public function onSubmit() {
-		if ($this->hasErrors())
-			return;
-		
-		$posting = new DB_Record();
-		$posting->board = $this->board;
-		$posting->user = Rakuun_User_Manager::getCurrentUser();
-		$posting->text = $this->text;
-		$posting->date = time();
-		Rakuun_DB_Containers::getBoardsPostingsContainer()->save($posting);
-		$this->text->setValue('');
-	}
-}
-
-?>
\ No newline at end of file
Index: source/intern/gui/panel/meta/board.php
===================================================================
--- source/intern/gui/panel/meta/board.php	(revision 0)
+++ source/intern/gui/panel/meta/board.php	(revision 0)
@@ -0,0 +1,35 @@
+<?php
+
+class Rakuun_Intern_GUI_Panel_Meta_Board extends Rakuun_Intern_GUI_Panel_Board {
+	public function __construct($name, $title = '') {
+		parent::__construct($name, Rakuun_DB_Containers::getBoardsMetaContainer()->getFilteredContainer(self::getMetaFilter()), $title);
+	}
+	
+	protected function getFilteredRecord() {
+		$record = new DB_Record();
+		$record->meta = Rakuun_User_Manager::getCurrentUser()->alliance->meta;
+		return $record;
+	}
+	
+	private static function getMetaFilter() {
+		$options = array();
+		$options['conditions'][] = array('meta = ?', Rakuun_User_Manager::getCurrentUser()->alliance->meta);
+		return $options; 
+	}
+	
+	protected function getPostingsContainer() {
+		return Rakuun_DB_Containers::getBoardsMetaPostingsContainer();
+	}
+	
+	protected function getVisitedContainer() {
+		return Rakuun_DB_Containers::getBoardsMetaLastVisitedContainer();
+	}
+	
+	public static function getNewPostingsCount(DB_Container $null = null, DB_Container $anothernull = null) {
+		return parent::getNewPostingsCount(
+			Rakuun_DB_Containers::getBoardsMetaContainer()->getFilteredContainer(self::getMetaFilter()),
+			Rakuun_DB_Containers::getBoardsMetaLastVisitedContainer()
+		);
+	}
+}
+?>
\ No newline at end of file
Index: source/intern/gui/panel/board/board.php
===================================================================
--- source/intern/gui/panel/board/board.php	(revision 450)
+++ source/intern/gui/panel/board/board.php	(working copy)
@@ -1,32 +1,146 @@
 <?php
 
 /**
- * Displays information of a single board.
+ * Displays all or one specific board.
  */
-class Rakuun_Intern_GUI_Panel_Board_Board extends GUI_Panel {
-	private $board = null;
-	
-	public function __construct($name, DB_Record $board) {
-		$this->board = $board;
-		parent::__construct($name);
+class Rakuun_Intern_GUI_Panel_Board extends GUI_Panel_PageView {
+	public function __construct($name, DB_Container $container = null, $title = '') {
+		$this->setItemsPerPage(10);
+		if ($container === null)
+			$container = Rakuun_DB_Containers::getBoardsContainer();
+			
+		parent::__construct($name, $container, $title);
 	}
 	
 	public function init() {
 		parent::init();
 		
 		$this->setTemplate(dirname(__FILE__).'/board.tpl');
-		$this->addPanel($link = new GUI_Control_Link('boardlink', $this->board->name, Router::get()->getCurrentModule()->getUrl(array('board' => $this->board->id))));
-		$options['conditions'][] = array('board = ?', $this->board);
+		if (($boardId = Router::get()->getCurrentModule()->getParam('board')) > 0) {
+			// Board darstellen
+			$board = $this->getContainer()->selectByIdFirst($boardId);
+			if ($board === null) {
+				$this->addError('Du hast keinen Zugriff auf das Forum mit der ID '.$boardId);
+				return;
+			}
+			$this->initWithSingleBoard($board);
+		} else {
+			$this->initWithBoards();
+		}
+	}
+	
+	protected function initWithSingleBoard(DB_Record $board) {
+		$this->addPanel(
+			new Rakuun_Intern_GUI_Panel_Board_PostingView(
+				'board',
+				$board,
+				$this->getPostingsContainer(),
+				$this->getVisitedContainer()
+			)
+		);
+	}
+	
+	private function initWithBoards() {
+		$user = Rakuun_User_Manager::getCurrentUser();
+		$options = array();
+		$options['order'] = 'date DESC';
+		$boards = $this->getContainer()->select($this->getOptions($options));
+		$this->addPanel($table = new GUI_Panel_Table('board'));
+		$table->addHeader(array('Name', 'Beiträge', 'Letzte Änderung'));
+		$module = Router::get()->getCurrentModule();
+		foreach ($boards as $board) {
+			$line = array();
+			$link = new GUI_Control_Link('boardlink', $board->name, $module->getUrl(array('board' => $board->id)));
+			$options = array();
+			$options['conditions'][] = array('user = ?', $user);
+			$options['conditions'][] = array('board = ?', $board);
+			$lastVisit = $this->getVisitedContainer()->selectFirst($options);
+			if (!$lastVisit || $lastVisit->date < $board->date)
+				$link->setAttribute('style', 'font-weight:bold');
+			$line[] = $link;
+			$line[] = $this->getPostingsContainer()->countByBoard($board);
+			$line[] = new GUI_Panel_Date('date'.$board->getPK(), $board->date);
+			$table->addLine($line);
+		}
+		$this->addPanel($blanko = new GUI_Panel('addboard'));
+		$blanko->addPanel($name = new GUI_Control_TextBox('name', null, 'Boardname'));
+		$name->addValidator(new GUI_Validator_Mandatory());
+		$name->addValidator(new GUI_Validator_RangeLength(2, 25));
+		$blanko->addPanel(new GUI_Control_SubmitButton('addboard', 'anlegen'));
+		$this->addPanel($blanko = new GUI_Panel('markread'));
+		$blanko->addPanel(new GUI_Control_SubmitButton('markread', 'Alle als gelesen markieren'));
+	}
+	
+	public function onAddBoard() {
+		if ($this->hasErrors())
+			return;
+		
+		$board = $this->getFilteredRecord();
+		$board->name = $this->addboard->name->getValue();
+		$board->date = time();
+		$this->getContainer()->save($board);
+		$this->addboard->name->resetValue();
+		$this->getModule()->invalidate();
+	}
+	
+	protected function getFilteredRecord() {
+		return new DB_Record();
+	}
+	
+	public static function getNewPostingsCount(DB_Container $boardsContainer = null, DB_Container $visitedContainer = null) {
+		if ($boardsContainer === null)
+			$boardsContainer = Rakuun_DB_Containers::getBoardsContainer();
+		if ($visitedContainer === null)
+			$visitedContainer = Rakuun_DB_Containers::getBoardsLastVisitedContainer();
+		$options = array();
 		$options['conditions'][] = array('user = ?', Rakuun_User_Manager::getCurrentUser());
-		$lastVisited = Rakuun_DB_Containers::getBoardsLastVisitedContainer()->selectFirst($options);
-		$postings = Rakuun_DB_Containers::getBoardsPostingsContainer();
-		$this->params->count = $postings->countByBoard($this->board);
-		$posting = $postings->selectByBoardFirst($this->board, array('order' => 'date DESC'));
-		if ($posting)
-			$this->addPanel(new GUI_Panel_Date('date', $posting->date));
-		if ($lastVisited === null || ($posting && ($posting->date > $lastVisited->date)))
-			$link->setAttribute('style', 'font-weight:bold');
+		$_visited = $visitedContainer->select($options);
+		$visited = array();
+		foreach ($_visited as $visit) {
+			$visited[$visit->board->getPK()] = $visit;
+		}
+		$boards = $boardsContainer->select();
+		$count = 0;
+		foreach ($boards as $board) {
+			if (isset($visited[$board->getPK()]) && $visited[$board->getPK()]->date < $board->date)
+				$count++;
+			if (!isset($visited[$board->getPK()]))
+				$count++;
+		}
+		return $count;
+	}
+	
+	public function onMarkRead() {
+		$user = Rakuun_User_Manager::getCurrentUser();
+		$lastVisitedContainer = $this->getVisitedContainer();
+		$boards = $this->getBoardsContainer()->select();
+		DB_Connection::get()->beginTransaction();
+		foreach ($boards as $board) {
+			$options = array();
+			$options['conditions'][] = array('board = ?', $board);
+			$options['conditions'][] = array('user = ?', $user);
+			$newVisit = $lastVisitedContainer->selectFirst($options);
+			if ($newVisit === null)
+				$newVisit = new DB_Record();
+			$newVisit->board = $board;
+			$newVisit->user = $user;
+			$newVisit->date = time();
+			$lastVisitedContainer->save($newVisit);
+		}
+		DB_Connection::get()->commit();
+		$this->getModule()->invalidate();
+	}
+	
+	protected function getBoardsContainer() {
+		return $this->getContainer();
+	}
+	
+	protected function getPostingsContainer() {
+		return Rakuun_DB_Containers::getBoardsPostingsContainer();
+	}
+	
+	protected function getVisitedContainer() {
+		return Rakuun_DB_Containers::getBoardsLastVisitedContainer();
 	}
 }
-
 ?>
\ No newline at end of file
Index: migrations/018.boards_update.php
===================================================================
--- migrations/018.boards_update.php	(revision 0)
+++ migrations/018.boards_update.php	(revision 0)
@@ -0,0 +1,146 @@
+<?php
+/**
+ * $1 - Creation of new tables and constraints
+ * $2 - Copy data from old to new tables
+ * $3 - Delete old data
+ * $4 - Remove unused columns
+ * $5 - Add new column
+ */
+
+// - $1 -- create new board-tables + constraints
+
+$queries[] = 'CREATE TABLE IF NOT EXISTS `boards_alliance` (
+  `id` mediumint(8) unsigned NOT NULL AUTO_INCREMENT,
+  `alliance` smallint(5) unsigned DEFAULT NULL,
+  `name` varchar(25) NOT NULL,
+  `date` int(10) unsigned NOT NULL,
+  PRIMARY KEY (`id`),
+  KEY `alliance` (`alliance`)
+) ENGINE=InnoDB  DEFAULT CHARSET=utf8;';
+
+$queries[] = 'ALTER TABLE `boards_alliance`
+  ADD CONSTRAINT `boards_alliance_ibfk_1` FOREIGN KEY (`alliance`) REFERENCES `alliances` (`id`) ON DELETE CASCADE;';
+
+$queries[] = 'CREATE TABLE IF NOT EXISTS `boards_alliance_postings` (
+  `id` mediumint(8) unsigned NOT NULL AUTO_INCREMENT,
+  `board` mediumint(8) unsigned DEFAULT NULL,
+  `user` int(10) unsigned DEFAULT NULL,
+  `text` text NOT NULL,
+  `date` int(10) unsigned NOT NULL,
+  `editdate` int(10) unsigned DEFAULT NULL,
+  PRIMARY KEY (`id`),
+  KEY `board` (`board`),
+  KEY `user` (`user`)
+) ENGINE=InnoDB  DEFAULT CHARSET=utf8;';
+
+$queries[] = 'ALTER TABLE `boards_alliance_postings`
+  ADD CONSTRAINT `boards_alliance_postings_ibfk_1` FOREIGN KEY (`board`) REFERENCES `boards_alliance` (`id`) ON DELETE CASCADE,
+  ADD CONSTRAINT `boards_alliance_postings_ibfk_2` FOREIGN KEY (`user`) REFERENCES `users` (`id`) ON DELETE SET NULL;';
+
+$queries[] = 'CREATE TABLE IF NOT EXISTS `boards_alliance_visited` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `board` mediumint(8) unsigned NOT NULL,
+  `user` int(10) unsigned NOT NULL,
+  `date` int(10) unsigned NOT NULL,
+  PRIMARY KEY (`id`),
+  KEY `board` (`board`),
+  KEY `user` (`user`)
+) ENGINE=InnoDB  DEFAULT CHARSET=utf8;';
+
+$queries[] = 'ALTER TABLE `boards_alliance_visited`
+  ADD CONSTRAINT `boards_alliance_visited_ibfk_1` FOREIGN KEY (`board`) REFERENCES `boards_alliance` (`id`) ON DELETE CASCADE,
+  ADD CONSTRAINT `boards_alliance_visited_ibfk_2` FOREIGN KEY (`user`) REFERENCES `users` (`id`) ON DELETE CASCADE;';
+
+$queries[] = 'CREATE TABLE IF NOT EXISTS `boards_meta` (
+  `id` mediumint(8) unsigned NOT NULL AUTO_INCREMENT,
+  `meta` smallint(5) unsigned DEFAULT NULL,
+  `name` varchar(25) NOT NULL,
+  `date` int(10) unsigned NOT NULL,
+  PRIMARY KEY (`id`),
+  KEY `meta` (`meta`)
+) ENGINE=InnoDB  DEFAULT CHARSET=utf8;';
+
+$queries[] = 'ALTER TABLE `boards_meta`
+  ADD CONSTRAINT `boards_meta_ibfk_1` FOREIGN KEY (`meta`) REFERENCES `metas` (`id`) ON DELETE CASCADE;';
+
+$queries[] = 'CREATE TABLE IF NOT EXISTS `boards_meta_postings` (
+  `id` mediumint(8) unsigned NOT NULL AUTO_INCREMENT,
+  `board` mediumint(8) unsigned DEFAULT NULL,
+  `user` int(10) unsigned DEFAULT NULL,
+  `text` text NOT NULL,
+  `date` int(10) unsigned NOT NULL,
+  `editdate` int(10) unsigned DEFAULT NULL,
+  PRIMARY KEY (`id`),
+  KEY `board` (`board`),
+  KEY `user` (`user`)
+) ENGINE=InnoDB  DEFAULT CHARSET=utf8;';
+
+$queries[] = 'ALTER TABLE `boards_meta_postings`
+  ADD CONSTRAINT `boards_meta_postings_ibfk_1` FOREIGN KEY (`board`) REFERENCES `boards_meta` (`id`) ON DELETE CASCADE,
+  ADD CONSTRAINT `boards_meta_postings_ibfk_2` FOREIGN KEY (`user`) REFERENCES `users` (`id`) ON DELETE SET NULL;';
+
+$queries[] = 'CREATE TABLE IF NOT EXISTS `boards_meta_visited` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `board` mediumint(8) unsigned NOT NULL,
+  `user` int(10) unsigned NOT NULL,
+  `date` int(10) unsigned NOT NULL,
+  PRIMARY KEY (`id`),
+  KEY `board` (`board`),
+  KEY `user` (`user`)
+) ENGINE=InnoDB  DEFAULT CHARSET=utf8;';
+
+$queries[] = 'ALTER TABLE `boards_meta_visited`
+  ADD CONSTRAINT `boards_meta_visited_ibfk_1` FOREIGN KEY (`board`) REFERENCES `boards_meta` (`id`) ON DELETE CASCADE,
+  ADD CONSTRAINT `boards_meta_visited_ibfk_2` FOREIGN KEY (`user`) REFERENCES `users` (`id`) ON DELETE CASCADE;';
+
+// - $2 -- copy old data to new tables
+
+$queries[] = 'INSERT INTO `boards_alliance` (`id`, `alliance`, `name`, `date`)
+	SELECT `boards`.`id`, `boards`.`alliance`, `boards`.`name`, '.time().'
+	FROM `boards`
+	WHERE `boards`.`alliance` is not null;';
+
+$queries[] = 'INSERT INTO `boards_meta` (`id`, `meta`, `name`, `date`)
+	SELECT `boards`.`id`, `boards`.`meta`, `boards`.`name`, '.time().'
+	FROM `boards`
+	WHERE `boards`.`meta` is not null;';
+
+$queries[] = 'INSERT INTO `boards_alliance_postings` (`board`, `user`, `text`, `date`, `editdate`)
+	SELECT `board`, `user`, `text`, `date`, `editdate`
+	FROM `boards_postings`
+	WHERE `boards_postings`.`board` IN (SELECT `id` FROM `boards_alliance`);';
+
+$queries[] = 'INSERT INTO `boards_meta_postings` (`board`, `user`, `text`, `date`, `editdate`)
+	SELECT `board`, `user`, `text`, `date`, `editdate`
+	FROM `boards_postings`
+	WHERE `boards_postings`.`board` IN (SELECT `id` FROM `boards_meta`);';
+
+// - $3 -- delete old data
+
+$queries[] = 'DELETE FROM `boards`
+	WHERE `id` in (SELECT `id` FROM `boards_alliance`);';
+
+$queries[] = 'DELETE FROM `boards`
+	WHERE `id` in (SELECT `id` FROM `boards_meta`);';
+
+$queries[] = 'DELETE FROM `boards_postings`
+	WHERE `board` in (SELECT `id` FROM `boards_alliance`);';
+
+$queries[] = 'DELETE FROM `boards_postings`
+	WHERE `board` in (SELECT `id` FROM `boards_meta`);';
+
+$queries[] = 'TRUNCATE TABLE `boards_visited`;';
+
+// - $4 -- remove unused columns and constraints
+
+$queries[] = 'ALTER TABLE `boards` DROP FOREIGN KEY `boards_ibfk_1`;';
+
+$queries[] = 'ALTER TABLE `boards` DROP FOREIGN KEY `boards_ibfk_2`;';
+
+$queries[] = 'ALTER TABLE `boards` DROP `alliance`, DROP `meta`, DROP `type`;'; 
+
+// - $5 -- add new column
+
+$queries[] = 'ALTER TABLE `boards` ADD `date` INT( 10 ) UNSIGNED NOT NULL;';
+
+?>
\ No newline at end of file
Index: source/intern/module/alliance/board.php
===================================================================
--- source/intern/module/alliance/board.php	(revision 429)
+++ source/intern/module/alliance/board.php	(working copy)
@@ -7,7 +7,7 @@
 		$this->setPageTitle('Forum - ['.$this->getUser()->alliance->tag.'] '.$this->getUser()->alliance->name);
 		$this->contentPanel->setTemplate(dirname(__FILE__).'/board.tpl');
 		
-		$this->contentPanel->addPanel(new Rakuun_GUI_Panel_Box('board', new Rakuun_Intern_GUI_Panel_Board_Boardview('Forum', Rakuun_Intern_GUI_Panel_Board_Boardview::TYPE_ALLIANCE), 'Allianz Forum'));
+		$this->contentPanel->addPanel(new Rakuun_GUI_Panel_Box('board', new Rakuun_Intern_GUI_Panel_Alliance_Board('Forum'), 'Allianz Forum'));
 	}
 	
 	// OVERRIDES / IMPLEMENTS --------------------------------------------------
Index: source/intern/gui/panel/board/boardview.tpl
===================================================================
--- source/intern/gui/panel/board/boardview.tpl	(revision 545)
+++ source/intern/gui/panel/board/boardview.tpl	(working copy)
@@ -1,27 +0,0 @@
-<? if($this->hasErrors()): ?>
-	<? $this->displayErrors() ?>
-<? else: ?>
-	<? if (isset($this->params->boardname)): ?>
-		<? $this->displayPanel($this->params->boardname); ?>
-	<? else: ?>
-		<? $this->displayPanel('markread'); ?>
-		<table summary="Forenübersicht" class="boardview">
-			<thead>
-				<tr>
-					<td>Name</td>
-					<td>Beiträge</td>
-					<td>Letzter Beitrag</td>
-				</tr>
-			</thead>
-			<tbody>
-				<? foreach ($this->panels as $panel): ?>
-					<? if (preg_match('=^board_(\d)+$=', $panel->getName())): ?>
-						<? $panel->display(); ?>
-					<? endif; ?>
-				<? endforeach; ?>
-			</tbody>
-		</table>
-		<hr />
-		<? $this->displayPanel('addboard'); ?>
-	<? endif; ?>
-<? endif; ?>
\ No newline at end of file
